t/helper/test-gvfs-protocol.c

```diff
@@ -0,0 +1,1739 @@
+		logerror("unable to send chunk");
+		return WORKER_RESULT__ERROR;
+	}
+
+	return WORKER_RESULT__OK;
+}
+
+static enum worker_result send_final_chunk(int fd)
+{
+	if (write_in_full(fd, "0\r\n\r\n", 5) < 0) {
+		logerror("unable to send final chunk");
+		return WORKER_RESULT__ERROR;
+	}
+
+	return WORKER_RESULT__OK;
+}
+
+static enum worker_result send_http_error(
+	int fd,
+	int http_code, const char *http_code_name,
+	int retry_after_seconds, enum worker_result wr)
+{
+	struct strbuf response_header = STRBUF_INIT;
+	struct strbuf response_content = STRBUF_INIT;
+	struct strbuf uuid = STRBUF_INIT;
+
+	strbuf_addf(&response_content, "Error: %d %s\r\n",
+		    http_code, http_code_name);
+	if (retry_after_seconds > 0)
+		strbuf_addf(&response_content, "Retry-After: %d\r\n",
+			    retry_after_seconds);
+
+	strbuf_addf  (&response_header, "HTTP/1.1 %d %s\r\n", http_code, http_code_name);
+	strbuf_addstr(&response_header, "Cache-Control: private\r\n");
+	strbuf_addstr(&response_header,	"Content-Type: text/plain\r\n");
+	strbuf_addf  (&response_header,	"Content-Length: %d\r\n", (int)response_content.len);
+	if (retry_after_seconds > 0)
+		strbuf_addf  (&response_header, "Retry-After: %d\r\n", retry_after_seconds);
+	strbuf_addf(  &response_header,	"Server: test-gvfs-protocol/%s\r\n", git_version_string);
+	strbuf_addf(  &response_header, "Date: %s\r\n", show_date(time(NULL), 0, DATE_MODE(RFC2822)));
+	gen_fake_uuid(&uuid);
+	strbuf_addf(  &response_header, "X-VSS-E2EID: %s\r\n", uuid.buf);
+	strbuf_addstr(&response_header, "\r\n");
+
+	if (write_in_full(fd, response_header.buf, response_header.len) < 0) {
+		logerror("unable to write response header");
+		wr = WORKER_RESULT__ERROR;
+		goto done;
+	}
+
+	if (write_in_full(fd, response_content.buf, response_content.len) < 0) {
+		logerror("unable to write response content body");
+		wr = WORKER_RESULT__ERROR;
+		goto done;
+	}
+
+done:
+	strbuf_release(&uuid);
+	strbuf_release(&response_header);
+	strbuf_release(&response_content);
+
+	return wr;
+}
+
+/*
+ * Return 1 if we send an AUTH error to the client.
+ */
+static int mayhem_try_auth(struct req *req, enum worker_result *wr)
+{
+	*wr = WORKER_RESULT__OK;
+
+	if (string_list_has_string(&mayhem_list, "http_401")) {
+		struct string_list_item *item;
+		int has_auth = 0;
+		for_each_string_list_item(item, &req->header_list) {
+			if (starts_with(item->string, "Authorization: Basic")) {
+				has_auth = 1;
+				break;
+			}
+		}
+		if (!has_auth) {
+			if (strstr(req->uri_base.buf, MY_SERVER_TYPE__ORIGIN)) {
+				logmayhem("http_401 (origin)");
+				*wr = send_http_error(1, 401, "Unauthorized", -1,
+						      WORKER_RESULT__OK);
+				return 1;
+			}
+
+			if (strstr(req->uri_base.buf, MY_SERVER_TYPE__CACHE)) {
+				logmayhem("http_400 (cacheserver)");
+				*wr = send_http_error(1, 400, "Bad Request", -1,
+						      WORKER_RESULT__OK);
+				return 1;
+			}
+
+			if (strstr(req->uri_base.buf, MY_SERVER_TYPE__ORIGIN)) {
```

Comment:
Will this code ever get hit?  This is the same check we're making above on line 302:

`if (strstr(req->uri_base.buf, MY_SERVER_TYPE__ORIGIN)) {`

And it always returns when the condition is true.

Reply:
good catch.  bit of a cut-n-paste error.  my intent for the 3rd case to be hit when no "server type" was given.